Multi structure

jcass11 opened the pull request:
Allows the mti_nma pipeline to be run on multiple structures by creating structure-named versions of every step.

JacksonMaxfield commented on .gitignore:
```diff
@@ -2,6 +2,7 @@
 /local_staging
 /data
 /dask-worker-space
+/~
```
This was caused by a bug and I _think_ I fixed it in one of my PRs to this branch. Can remove.

JacksonMaxfield commented on README.md:
````diff
@@ -160,13 +163,13 @@ The following commands are the minimal set of commands to get setup for running
 
     ```bash
-    mti_nma all run --path_blender /allen/aics/modeling/jacksonb/applications/blender-2.82-linux64/blender --nsamples {int} --distributed
+    mti_nma all run --path_blender /allen/aics/modeling/jacksonb/applications/blender-2.82-linux64/blender --nsamples {int} --distributed True
````
You actually don't need the `True` at the end. Simply by passing `--distributed` it should assume True

JacksonMaxfield commented on README.md:
````diff
@@ -160,13 +163,13 @@ The following commands are the minimal set of commands to get setup for running
 
     ```bash
-    ssh -A -J slurm-master -L {port_provided_by_log}:{node_you_are_on}:{port_provided_by_log} {node_you_are_on}
+    ssh -A -J <username>@slurm-master.corp.alleninstitute.org -L {port_provided_by_log}:{node_you_are_on}:{port_provided_by_log} <username>@{node_you_are_on}
````
Lol nice WFH :upside_down_face:

JacksonMaxfield commented on mti_nma/bin/all.py:
```diff
@@ -63,37 +66,50 @@ def run(
                 # Do not include ms
                 log_dir_name = datetime.now().isoformat().split(".")[0]
-                log_dir = Path(f"~/.dask_logs/mti_nma/{log_dir_name}")
+                log_dir = Path(f"~/.dask_logs/mti_nma/{log_dir_name}").expanduser()
```
Yep this `.expanduser()` solves the `~` directory from being made. So can remove the gitignored line.

JacksonMaxfield commented on mti_nma/bin/all.py:
```diff
@@ -102,9 +118,10 @@ def run(
-                # Scale workers
-                cluster.adapt(minimum_jobs=1, maximum_jobs=40)
+                # Set adaptive worker settings
+                cluster.scale_up(40)
```
I would change this to
```suggestion
                cluster.adapt(minimum_jobs=30, maximum_jobs=60)
```

JacksonMaxfield commented on mti_nma/bin/all.py:
```diff
@@ -132,49 +151,83 @@ def run(
+
+            # Run flow, get ending state, and visualize pipeline
+            flow.run(executor=exe)
+            flow.visualize()
```
May want to remove this line or make it a parameter `--viz`. the dependencies required for `flow.visualize()` can't just be bare installed with `pip`.

JacksonMaxfield commented on mti_nma/steps/compare_nuc_cell/compare_nuc_cell.py:
```diff
@@ -0,0 +1,125 @@
+        debug: bool
+            A debug flag for the developer to use to manipulate how much data runs,
+            how it is processed, etc.
+            Default: False (Do not debug)
```
I should have told you this so my bad, but these should be folded into the `Parameters` section of each docstring where needed. They are just always passed by default so if you don't use them then get rid of them but if you do put them in the primary `Parameters` sections.

JacksonMaxfield commented on mti_nma/steps/nma/nma.py:
```diff
@@ -90,6 +89,9 @@ def run(
         """
 
+        # Set upstream task using structure-named version of previous step
+        self.direct_upstream_tasks = [Avgshape(step_name=f"avgshape_{struct}")]
```
I don't think this will work...

jcass11 replied:
Hm - then how am I supposed to use the structure-named (non-generic) upstream tasks so that pull will work?

JacksonMaxfield replied:
Right, this is why I think `pull` is currently broken. This is set in `run`. Not `init`. The way step works is that when you want to do any operation. You init the class then run it there is no statefulness between commands.

jcass11 replied:
Okay I understand - thanks

JacksonMaxfield commented on mti_nma/steps/shparam/shparam.py:
```diff
@@ -140,6 +143,9 @@ def run(
         """
 
+        # Set upstream task using structure-named version of previous step
+        self.direct_upstream_tasks = [Singlecell(step_name=f"singlecell_{struct}")]
```
Similar here, don't think this will work.

JacksonMaxfield commented on setup.py:
```diff
@@ -64,6 +63,8 @@
     "scikit-image",
     "seaborn",
     "vtk",
+    "graphviz",
```
Does this fully pip install? huh. I really thought it had to be conda.

jcass11 commented:
> Should be able to `git rm ~` or similar. Sorry about that. When I said I had fixed it I meant I had fixed the problem of it creating more files but should have also mentioned to delete those files as well.
> 
> Thanks for making the changes and nice work!

When I try `git rm ~` I get `fatal: /home/juliec: '/home/juliec' is outside repository` which makes me very nervous about messing around with this....

jcass11 commented:
> > Should be able to `git rm ~` or similar. Sorry about that. When I said I had fixed it I meant I had fixed the problem of it creating more files but should have also mentioned to delete those files as well.
> > Thanks for making the changes and nice work!
> 
> When I try `git rm ~` I get `fatal: /home/juliec: '/home/juliec' is outside repository` which makes me very nervous about messing around with this....

For our records, @JacksonMaxfield let me know that:

> It was rm -Rf \~ to escape the normal behavior and look in the current directory for "~" dir

and pushed a commit to address this.

vianamp commented on mti_nma/steps/avgshape/avgshape.py:
```diff
@@ -108,6 +106,6 @@ def run(self, sh_df=None, struct="Nuc", **kwargs):
         # Save manifest as csv
         self.manifest.to_csv(
-            self.step_local_staging_dir / Path(f"manifest_{struct}.csv"), index=False
+            self.step_local_staging_dir / Path(f"manifest.csv"), index=False
```
I like this :)

vianamp commented on setup.py:
```diff
@@ -42,14 +42,13 @@
-    "aicsimageio",
-    "aicsshparam",
+    "aicsimageio>=3.1.2",
+    "aicsshparam>=0.0.7",
```
Please update this to >=0.0.9.
